File: C++/findDisappearedNumbers.cpp
/**
 * @File    : findDisappearedNumbers.cpp
 * @Brief   : 找到所有数组中消失的数字  
 * @Link    : https://leetcode-cn.com/problems/find-all-numbers-disappeared-in-an-array/
 * @Author  : Wei Li
 * @Date    : 2021-07-23
*/

#include <iostream>
#include <vector>

class Solution
{
public:
    std::vector<int> findDisappearedNumbers(std::vector<int> &nums)
    {
        std::vector<int> ans;
        for (const int &num : nums)
        {
            int pos = abs(num) - 1;
            if (nums[pos] > 0)
            {
                nums[pos] = -nums[pos];
            }
        }
        for (int i = 0; i < nums.size(); ++i)
        {
            if (nums[i] > 0)
            {
                ans.push_back(i + 1);
            }
        }
        return ans;
    }
};

template <typename T>
void DisplayElement(const T& container)
{
    for (auto element : container)
    {
        std::cout << element << " ";
    }
    std::cout << std::endl;
}

// --------------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums{4, 3, 2, 7, 8, 2, 3, 1};

    Solution solution;
    auto disappeared_vec = solution.findDisappearedNumbers(nums);
    DisplayElement(disappeared_vec);

    return 0;
}

File: Python/findDisappearedNumbers.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: findDisappearedNumbers.py
@Function: 找到所有数组中消失的数字
@Link: https://leetcode-cn.com/problems/find-all-numbers-disappeared-in-an-array/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-23
"""

from typing import List


class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums)
        for num in nums:
            x = (num - 1) % n
            nums[x] += n

        ret = [i + 1 for i, num in enumerate(nums) if num <= n]
        return ret


# --------------------------------------
if __name__ == "__main__":
    nums = [4, 3, 2, 7, 8, 2, 3, 1]

    solution = Solution()
    disappeared_vec = solution.findDisappearedNumbers(nums)
    print(f"The solution of this problem is : {disappeared_vec}")
